test/blackbox-tests/test-cases/pkg/source-stanza/mixed-opam-dune.t

```diff
@@ -0,0 +1,34 @@
+We try to use a project that has both opam files and a dune-project file. We
+should favor the dune metadata in such a case.
+
+  $ . ../helpers.sh
+
+  $ mkrepo
+  $ add_mock_repo_if_needed
+
+  $ cat >dune-project <<EOF
+  > (lang dune 3.13)
+  > (package_source
+  >  (url "file://$PWD/_source")
+  >  (package (name foo))
+  >  (package (name bar)))
+  > (package
+  >  (name main)
+  >  (depends foo bar))
+  > EOF
+
+  $ mkdir _source
+  $ cat >_source/dune-project <<EOF
+  > (lang dune 3.13)
+  > (package (name foo))
+  > (package (name bar))
+  > EOF
+  $ cat >_source/bar.opam <<EOF
+  > opam-version: "2.0"
+  > build: [ "echo" "bar" ]
+  > EOF
+
+  $ dune pkg lock
+  Solution for dune.lock:
+  - bar.dev
+  - foo.dev
```

Comment:
> or set version: X.Y in bar.opam and show that bar.dev instead of bar.X.Y was chosen.

According to the semantics of pins, the version in the `bar.opam` would be ignored in favor of the version set by the caller. So I did your first suggestion.